distributed/shuffle/_rechunk.py

```diff
@@ -216,27 +228,216 @@ def rechunk_p2p(x: da.Array, chunks: ChunkedAxes) -> da.Array:
+    token: str
+    chunks: ChunkedAxes
+    chunks_input: ChunkedAxes
+    name_input: str
+    disk: bool
+    keepmap: np.ndarray
+
+    _cached_dict: _T_LowLevelGraph | None
+
+    def __init__(
+        self,
+        name: str,
+        token: str,
+        chunks: ChunkedAxes,
+        chunks_input: ChunkedAxes,
+        name_input: str,
+        disk: bool,
+        keepmap: np.ndarray | None = None,
+        annotations: Mapping[str, Any] | None = None,
+    ):
+        import numpy as np
+
+        self.name = name
+        self.token = token
+        self.chunks = chunks
+        self.chunks_input = chunks_input
+        self.name_input = name_input
+        self.disk = disk
+        if keepmap is not None:
+            self.keepmap = keepmap
+        else:
+            shape = tuple(len(axis) for axis in chunks)
+            self.keepmap = np.ones(shape, dtype=bool)
+        self._cached_dict = None
+        super().__init__(annotations=annotations)
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}<name='{self.name}', chunks={self.chunks}>"
+
+    def get_output_keys(self) -> set[Key]:
+        import numpy as np
+
+        return {
+            (self.name,) + nindex
+            for nindex in np.ndindex(tuple(len(axis) for axis in self.chunks))
+            if self.keepmap[nindex]
+        }
+
+    def is_materialized(self) -> bool:
+        return self._cached_dict is not None
+
+    @property
+    def _dict(self) -> _T_LowLevelGraph:
+        """Materialize full dict representation"""
+        dsk: _T_LowLevelGraph
+        if self._cached_dict is not None:
+            return self._cached_dict
+        else:
+            dsk = self._construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def __getitem__(self, key: Key) -> tuple:
+        return self._dict[key]
+
+    def __iter__(self) -> Iterator[Key]:
+        return iter(self._dict)
+
+    def __len__(self) -> int:
+        return len(self._dict)
+
+    def _cull(self, keepmap: np.ndarray) -> P2PRechunkLayer:
+        return P2PRechunkLayer(
+            name=self.name,
+            token=self.token,
+            chunks=self.chunks,
+            chunks_input=self.chunks_input,
+            name_input=self.name_input,
+            disk=self.disk,
+            keepmap=keepmap,
+            annotations=self.annotations,
+        )
+
+    def _keys_to_indices(self, keys: Iterable[Key]) -> set[tuple[int, ...]]:
+        """Simple utility to convert keys to chunk indices."""
+        chunks = set()
+        for key in keys:
+            if not isinstance(key, tuple) or len(key) < 2 or key[0] != self.name:
+                continue
+            chunk = cast(tuple[int, ...], key[1:])
+            assert all(isinstance(index, int) for index in chunk)
+            chunks.add(chunk)
+        return chunks
+
+    def cull(
+        self, keys: set[Key], all_keys: Collection[Key]
+    ) -> tuple[P2PRechunkLayer, dict]:
```

Comment:
I don't love this... it feels like we're doing the same work twice. I know that the culling API forces us to generate this dict but it is still odd.

Reply:
Agreed, I'm still trying to figure out if there's a better way to avoid work here.

Reply:
I'm also worried about how this will look once we move to expressions where we no longer have the culling as such a concept. However, array expressions are in such an early stage of development that this hasn't been dealt with, yet, from what I can tell.

@phofl any insights by any chance (might also be easier to hop on a call)

Reply:
yeah call makes sense